packages/api/cms-api/src/dam/files/file-warning.service.ts

```diff
@@ -0,0 +1,90 @@
+import { Inject, Injectable } from "@nestjs/common";
+
+import { CreateWarningsServiceInterface } from "../../warnings/decorators/create-warnings.decorator";
+import { CreateWarningInput } from "../../warnings/dto/create-warning.input";
+import { DamConfig } from "../dam.config";
+import { DAM_CONFIG } from "../dam.constants";
+import { FileInterface } from "./entities/file.entity";
+import { LicenseType } from "./entities/license.embeddable";
+
+@Injectable()
+export class FileWarningService implements CreateWarningsServiceInterface<FileInterface> {
+    constructor(
+        @Inject(DAM_CONFIG) private readonly config: DamConfig,
+        private readonly entityManager: EntityManager,
+    ) {}
+
+    async *bulkCreateWarnings() {
+        if (!this.config.enableLicenseFeature) return; // license feature not enabled, no warnings
+
+        const soonToExpireDate = new Date();
+        soonToExpireDate.setDate(soonToExpireDate.getDate() + 30);
+
+        const filterQuery: FilterQuery<FileInterface> = [
+            {
+                license: {
+                    durationTo: {
+                        $lt: soonToExpireDate,
+                    },
+                },
+            },
+        ];
+        if (this.config.requireLicense) {
+            filterQuery.push({ license: null });
+        }
+
+        let files = [];
+        let offset = 0;
+        const limit = 50;
+        do {
+            files = await this.entityManager.getRepository<FileInterface>("DamFile").find(
+                {
+                    $or: filterQuery,
+                },
+                { limit, offset },
+            );
+
+            for (const file of files) {
+                const warnings = await this.createWarnings(file);
+                yield { warnings, tableRowId: file.id };
+            }
+            offset += limit;
+        } while (files.length > 0);
+    }
+
+    async createWarnings(entity: FileInterface) {
+        if (!this.config.enableLicenseFeature) return []; // license feature not enabled, no warnings
+
+        const warnings: CreateWarningInput[] = [];
+
+        if (entity.license?.durationTo) {
+            const soonToExpireDate = new Date();
+            soonToExpireDate.setDate(soonToExpireDate.getDate() + 30);
+
+            if (entity.license.durationTo < new Date()) {
+                warnings.push({
+                    severity: "high",
+                    message: "fileLicenseExpired",
+                });
+            } else if (entity.license.durationTo < soonToExpireDate) {
+                warnings.push({
+                    severity: "medium",
+                    message: "fileLicenseSoonToExpire",
+                });
+            }
+        }
+
+        // if license feature is required, check if licenses are set
+        if (this.config.requireLicense) {
+            if (!entity.license?.durationTo && entity.license?.type !== LicenseType.ROYALTY_FREE) {
```

Comment:
Why do you check for the duration here? To see if a license has expired?

Reply:
This check is only needed when requireLicense is activated. I want to verify if there are any entries in the database without a license.

This situation should only occur if the project initially didn’t require a license but later made it mandatory.

I could also check if type is undefined, or go further and check if any of type, durationFrom, or durationTo are undefined. Ideally, either all three should be set or none at all.

I initially focused on durationTo since this seems to be the "important" field for the license in my opinion. But on second thought, it’s probably better to check all three to ensure the license is properly set. What do you think?

Reply:
Thanks for the explanation! It's probably sufficient to check for the license type since it's required in the Admin: https://github.com/vivid-planet/comet/blob/next/packages/admin/cms-admin/src/dam/FileForm/FileSettingsFields.tsx#L192

Reply:
Maybe we could introduce a variable `hasLicense` to make it more clear.

Reply:
Good idea. I've introduced a variable `isLicenseMissing` instead though to avoid double negation
3ccd3c4065fcda02ae0523131b89f6ff03b5f2f6